File: src/sources/console/format.ts
import type { LogPart, TextPart, TextStyle, ValueNode } from '../../core/types.js';

/** Captures a value for display. */
export type ValueCapture = (value: unknown) => ValueNode;

const SAFE_COLOR_FUNCTION = /^(rgba?|hsla?|hwb|lab|lch|oklab|oklch|color)\([\d\s.,%+\-/a-z]*\)$/i;
const SAFE_COLOR_HEX = /^#[0-9a-f]{3,8}$/i;
const SAFE_COLOR_NAME = /^[a-z]{3,30}$/i;
const SPECIFIER = /%([sdifoOc%])/g;

/** Returns whether a CSS color can be used as is. Anything that could load a resource is rejected. */
export const isSafeColor = (value: string): boolean => {
	const color = value.trim();

	return (
		SAFE_COLOR_HEX.test(color) || SAFE_COLOR_FUNCTION.test(color) || SAFE_COLOR_NAME.test(color)
	);
};

/** Splits a CSS value on spaces that are not inside parentheses. */
const splitCssValue = (value: string): string[] => {
	const tokens: string[] = [];
	let depth = 0;
	let current = '';

	for (const character of value) {
		if (character === '(') {
			depth++;
		} else if (character === ')') {
			depth = Math.max(0, depth - 1);
		}

		if (/\s/.test(character) && depth === 0) {
			if (current) {
				tokens.push(current);
				current = '';
			}
		} else {
			current += character;
		}
	}

	if (current) {
		tokens.push(current);
	}

	return tokens;
};

/**
 * Reads the CSS given to `%c` and keeps only what the viewer can draw: text and background
 * color, weight, style and decoration. Everything else is ignored, including any value that
 * refers to a URL, so a log message cannot make the page load anything.
 */
export const parseConsoleCss = (css: string): TextStyle | undefined => {
	const style: TextStyle = {};

	for (const declaration of css.split(';')) {
		const separator = declaration.indexOf(':');

		if (separator < 0) {
			continue;
		}

		const property = declaration.slice(0, separator).trim().toLowerCase();
		const value = declaration.slice(separator + 1).trim();

		if (property === 'color' && isSafeColor(value)) {
			style.color = value;
		} else if (property === 'background' || property === 'background-color') {
			const color = splitCssValue(value).find((token) => isSafeColor(token));

			if (color) {
				style.background = color;
			}
		} else if (property === 'font-weight') {
			const weight = Number(value);

			if (value === 'bold' || value === 'bolder' || weight >= 600) {
				style.bold = true;
			}
		} else if (property === 'font-style' && /^(italic|oblique)/i.test(value)) {
			style.italic = true;
		} else if (property === 'text-decoration' || property === 'text-decoration-line') {
			if (/underline/i.test(value)) {
				style.underline = true;
			}

			if (/line-through/i.test(value)) {
				style.strikethrough = true;
			}
		}
	}

	return Object.keys(style).length > 0 ? style : undefined;
};

const toText = (value: unknown): string => {
	if (typeof value === 'string') {
		return value;
	}

	try {
		return String(value);
	} catch {
		return Object.prototype.toString.call(value);
	}
};

const toInteger = (value: unknown): string => {
	return typeof value === 'symbol' ? 'NaN' : String(parseInt(toText(value), 10));
};

const toFloat = (value: unknown): string => {
	return typeof value === 'symbol' ? 'NaN' : String(parseFloat(toText(value)));
};

/**
 * Applies the format specifiers in the first argument, following the Formatter operation of
 * the Console Standard, and returns the parts of the message together with the arguments the
 * specifiers did not consume.
 *
 * `%s` converts with `String`, `%d` and `%i` with `parseInt`, `%f` with `parseFloat`, `%o`
 * and `%O` insert the value itself, and `%c` styles the text that follows it. `%%` writes a
 * percent sign. A specifier with no argument left stays in the text as written.
 */
export const applyFormat = (
	args: readonly unknown[],
	capture: ValueCapture
): { parts: LogPart[]; rest: unknown[] } => {
	const [format, ...values] = args;
	const parts: LogPart[] = [];

	if (typeof format !== 'string') {
		return { parts, rest: [...args] };
	}

	let style: TextStyle | undefined;
	let text = '';
	let last = 0;
	let consumed = 0;

	const flush = (): void => {
		if (text) {
			const part: TextPart = { type: 'text', text };

			if (style) {
				part.style = style;
			}

			parts.push(part);
			text = '';
		}
	};

	for (const match of format.matchAll(SPECIFIER)) {
		const specifier = match[1];
		const index = match.index ?? 0;

		text += format.slice(last, index);
		last = index + match[0].length;

		if (specifier === '%') {
			text += '%';
			continue;
		}

		if (consumed >= values.length) {
			text += match[0];
			continue;
		}

		const value = values[consumed++];

		switch (specifier) {
			case 's':
				text += toText(value);
				break;
			case 'd':
			case 'i':
				text += toInteger(value);
				break;
			case 'f':
				text += toFloat(value);
				break;
			case 'c':
				flush();
				style = parseConsoleCss(toText(value));
				break;
			default:
				flush();
				parts.push({ type: 'value', value: capture(value) });
				break;
		}
	}

	text += format.slice(last);
	flush();

	return { parts, rest: values.slice(consumed) };
};

/**
 * Turns the arguments of a console call into the parts of an entry.
 *
 * With more than one argument and a string first, the format specifiers are applied. The
 * remaining arguments follow, separated by spaces: strings as plain text and everything else
 * as a captured value.
 */
export const formatArguments = (args: readonly unknown[], capture: ValueCapture): LogPart[] => {
	let parts: LogPart[] = [];
	let rest: unknown[] = [...args];

	if (args.length > 1 && typeof args[0] === 'string') {
		const formatted = applyFormat(args, capture);

		parts = formatted.parts;
		rest = formatted.rest;
	}

	for (const value of rest) {
		if (parts.length > 0) {
			parts.push({ type: 'text', text: ' ' });
		}

		parts.push(
			typeof value === 'string'
				? { type: 'text', text: value }
				: { type: 'value', value: capture(value) }
		);
	}

	return parts;
};

File: packages/flutter/lib/src/sources/console/format.dart
import 'package:lognal/src/core/types.dart';
import 'package:lognal/src/sources/console/css_color.dart';

/// Captures a value for display.
typedef ValueCapture = ValueNode Function(Object? value);

final RegExp _specifier = RegExp('%([sdifoOc%])');

/// Splits a CSS value on spaces that are not inside parentheses.
List<String> _splitCssValue(String value) {
  final List<String> tokens = <String>[];
  final StringBuffer current = StringBuffer();
  int depth = 0;

  for (final String character in value.split('')) {
    if (character == '(') {
      depth++;
    } else if (character == ')') {
      depth = depth > 0 ? depth - 1 : 0;
    }

    if (character.trim().isEmpty && depth == 0) {
      if (current.isNotEmpty) {
        tokens.add(current.toString());
        current.clear();
      }
    } else {
      current.write(character);
    }
  }

  if (current.isNotEmpty) {
    tokens.add(current.toString());
  }

  return tokens;
}

/// Reads the CSS given to `%c` and keeps only what the viewer can draw: text and
/// background color, weight, style and decoration. Everything else is ignored,
/// including any value that refers to a URL, so a log message cannot make the
/// application load anything.
LogTextStyle? parseConsoleCss(String css) {
  int? color;
  int? background;
  bool bold = false;
  bool italic = false;
  bool underline = false;
  bool strikethrough = false;

  for (final String declaration in css.split(';')) {
    final int separator = declaration.indexOf(':');

    if (separator < 0) {
      continue;
    }

    final String property = declaration.substring(0, separator).trim().toLowerCase();
    final String value = declaration.substring(separator + 1).trim();

    if (property == 'color') {
      color = parseCssColor(value);
    } else if (property == 'background' || property == 'background-color') {
      for (final String token in _splitCssValue(value)) {
        final int? parsed = parseCssColor(token);

        if (parsed != null) {
          background = parsed;
          break;
        }
      }
    } else if (property == 'font-weight') {
      final double? weight = double.tryParse(value);

      if (value == 'bold' || value == 'bolder' || (weight != null && weight >= 600)) {
        bold = true;
      }
    } else if (property == 'font-style' && RegExp('^(italic|oblique)').hasMatch(value)) {
      italic = true;
    } else if (property == 'text-decoration' || property == 'text-decoration-line') {
      if (value.contains('underline')) {
        underline = true;
      }

      if (value.contains('line-through')) {
        strikethrough = true;
      }
    }
  }

  final LogTextStyle style = LogTextStyle(
    color: color == null ? null : RgbTextColor(color),
    background: background == null ? null : RgbTextColor(background),
    bold: bold,
    italic: italic,
    underline: underline,
    strikethrough: strikethrough,
  );

  return style.isEmpty ? null : style;
}

String _toText(Object? value) => value == null ? 'null' : value.toString();

String _toInteger(Object? value) {
  if (value is num) {
    return '${value.truncate()}';
  }

  final String text = _toText(value);
  final RegExpMatch? digits = RegExp(r'^\s*[+-]?\d+').firstMatch(text);

  return digits == null ? 'NaN' : '${int.parse(digits.group(0)!.trim())}';
}

String _toFloat(Object? value) {
  if (value is num) {
    return '${value.toDouble()}';
  }

  final String text = _toText(value);
  final RegExpMatch? number = RegExp(r'^\s*[+-]?\d*\.?\d+(?:[eE][+-]?\d+)?').firstMatch(text);

  return number == null ? 'NaN' : '${double.parse(number.group(0)!.trim())}';
}

/// The parts of a formatted message, and the arguments no specifier consumed.
class FormattedMessage {
  /// Creates the result.
  const FormattedMessage(this.parts, this.rest);

  /// The parts the format string produced.
  final List<LogPart> parts;

  /// The arguments the specifiers did not consume.
  final List<Object?> rest;
}

/// Applies the format specifiers in the first argument, and returns the parts of
/// the message together with the arguments the specifiers did not consume.
///
/// `%s` converts with `toString`, `%d` and `%i` to an integer, `%f` to a double,
/// `%o` and `%O` insert the value itself, and `%c` styles the text that follows
/// it. `%%` writes a percent sign. A specifier with no argument left stays in
/// the text as written.
FormattedMessage applyFormat(List<Object?> args, ValueCapture capture) {
  final List<LogPart> parts = <LogPart>[];

  if (args.isEmpty || args.first is! String) {
    return FormattedMessage(parts, List<Object?>.of(args));
  }

  final String format = args.first as String;
  final List<Object?> values = args.sublist(1);
  LogTextStyle? style;
  StringBuffer text = StringBuffer();
  int last = 0;
  int consumed = 0;

  void flush() {
    if (text.isNotEmpty) {
      parts.add(TextPart(text.toString(), style: style));
      text = StringBuffer();
    }
  }

  for (final RegExpMatch match in _specifier.allMatches(format)) {
    final String specifier = match.group(1)!;

    text.write(format.substring(last, match.start));
    last = match.end;

    if (specifier == '%') {
      text.write('%');
      continue;
    }

    if (consumed >= values.length) {
      text.write(match.group(0));
      continue;
    }

    final Object? value = values[consumed++];

    switch (specifier) {
      case 's':
        text.write(_toText(value));
      case 'd':
      case 'i':
        text.write(_toInteger(value));
      case 'f':
        text.write(_toFloat(value));
      case 'c':
        flush();
        style = parseConsoleCss(_toText(value));
      default:
        flush();
        parts.add(ValuePart(capture(value)));
    }
  }

  text.write(format.substring(last));
  flush();

  return FormattedMessage(parts, values.sublist(consumed));
}

/// Turns the arguments of a logging call into the parts of an entry.
///
/// With more than one argument and a string first, the format specifiers are
/// applied. The remaining arguments follow, separated by spaces: strings as
/// plain text and everything else as a captured value.
List<LogPart> formatArguments(List<Object?> args, ValueCapture capture) {
  List<LogPart> parts = <LogPart>[];
  List<Object?> rest = List<Object?>.of(args);

  if (args.length > 1 && args.first is String) {
    final FormattedMessage formatted = applyFormat(args, capture);

    parts = formatted.parts;
    rest = formatted.rest;
  }

  for (final Object? value in rest) {
    if (parts.isNotEmpty) {
      parts.add(const TextPart(' '));
    }

    parts.add(value is String ? TextPart(value) : ValuePart(capture(value)));
  }

  return parts;
}
